From be4f5b3d719698030734cfb9719711f212a4613d Mon Sep 17 00:00:00 2001
From: handsomeyingyan <handsomeyingyan@github.com>
Date: Tue, 13 Oct 2020 10:29:45 +0800
Subject: [PATCH] fix sun8i v3s simple framebuffer

---
 arch/arm/boot/dts/sun8i-v3s.dtsi | 17 +++++++++++++++++
 1 file changed, 17 insertions(+)

diff --git a/arch/arm/boot/dts/sun8i-v3s.dtsi b/arch/arm/boot/dts/sun8i-v3s.dtsi
index 3979394b5..007d3074a 100644
--- a/arch/arm/boot/dts/sun8i-v3s.dtsi
+++ b/arch/arm/boot/dts/sun8i-v3s.dtsi
@@ -44,12 +44,29 @@
 #include <dt-bindings/clock/sun8i-v3s-ccu.h>
 #include <dt-bindings/reset/sun8i-v3s-ccu.h>
 #include <dt-bindings/pwm/pwm.h>
+#include <dt-bindings/clock/sun8i-de2.h>
+
 
 / {
     #address-cells = <1>;
     #size-cells = <1>;
     interrupt-parent = <&gic>;
 
+	chosen {
+		#address-cells = <1>;
+		#size-cells = <1>;
+		ranges;
+
+		framebuffer-lcd {
+			compatible = "allwinner,simple-framebuffer",
+				     "simple-framebuffer";
+			allwinner,pipeline = "mixer0-lcd0";
+			clocks = <&display_clocks CLK_MIXER0>,
+				 <&ccu CLK_TCON0>;
+			status = "disabled";
+		};
+	};
+
     cpus {
         #address-cells = <1>;
         #size-cells = <0>;
-- 
2.25.1

