Index: linux-4.9.37/scripts/Makefile
===================================================================
--- linux-4.9.37.orig/scripts/Makefile
+++ linux-4.9.37/scripts/Makefile
@@ -24,7 +24,7 @@ hostprogs-$(CONFIG_SYSTEM_EXTRA_CERTIFIC
 HOSTCFLAGS_sortextable.o = -I$(srctree)/tools/include
 HOSTCFLAGS_asn1_compiler.o = -I$(srctree)/include
 HOSTLOADLIBES_sign-file = -lcrypto
-HOSTLOADLIBES_extract-cert = -lcrypto
+HOSTLOADLIBES_extract-cert = -lcrypto -lpthread
 
 always		:= $(hostprogs-y) $(hostprogs-m)
 
