From 7a88c17ffceb4db5dd85f952aeff59f000ed252c Mon Sep 17 00:00:00 2001
From: Clark Williams <williams@redhat.com>
Date: Thu, 27 Aug 2020 17:51:48 -0500
Subject: [PATCH 507/507] Linux 4.14.195-rt94 REBASE

Signed-off-by: Clark Williams <williams@redhat.com>
---
 localversion-rt | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/localversion-rt b/localversion-rt
index 90290c642ed5..8d02a9bac500 100644
--- a/localversion-rt
+++ b/localversion-rt
@@ -1 +1 @@
--rt29
+-rt94
-- 
2.26.2

