--- a/arch/arm/boot/dts/suniv-f1c100s-licheepi-nano.dts
+++ b/arch/arm/boot/dts/suniv-f1c100s-licheepi-nano.dts
@@ -79,6 +79,14 @@
 	status = "okay";
 };
 
+&cedarx {
+	status = "okay";
+};
+
+&ion {
+	status = "okay";
+};
+
 &spi0 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&spi0_pins_a>;
@@ -87,14 +95,55 @@
 	flash@0 {
 		#address-cells = <1>;
 		#size-cells = <1>;
-		compatible = "winbond,w25q128", "jedec,spi-nor";
-		/*		compatible = "spin-nand", "spi-nand";*/
+		/*compatible = "winbond,w25q128", "jedec,spi-nor";*/
+		compatible = "spin-nand", "spi-nand";
 		/*compatible = "linux,spidev";		*/
 		reg = <0>;
-		spi-max-frequency = <40000000>;
+		spi-max-frequency = <100000000>;
+		spi-rx-bus-width = <2>;
+		spi-tx-bus-width = <2>;
+		partitions {
+			compatible = "fixed-partitions";
+			#address-cells = <1>;
+			#size-cells = <1>;
+
+			partition@0 {
+				label = "spl";
+				reg = <0x000000 0x0080000>;
+			};
+			partition@1 {
+				label = "u-boot";
+				reg = <0x080000 0x0100000>;
+			};
+			partition@2 {
+				label = "dtb";
+				reg = <0x180000 0x0040000>;
+			};
+			partition@3 {
+				label = "kernel";
+				reg = <0x1c0000 0x0600000>;
+			};
+			partition@4 {
+				label = "rootfs";
+				reg = <0x7c0000 0x77c0000>;
+			};
+		};
+	};
+};
+
+&twi0 {
+	pinctrl-names = "default";
+	pinctrl-0 = <&twi0_pins>;
+	status = "okay";
+	ts: twi0@0x48 {
+		compatible = "nsiway,ns2009";
+		i2c-max-frequency = <100000>;
+		reg = <0x48>;
+		status = "okay";
 	};
 };
 
+
 &tcon0 {
 	pinctrl-names = "default";
 	pinctrl-0 = <&lcd_rgb565_pins>;
@@ -118,9 +167,17 @@
 	status = "disable";
 };
 
+&pio {
+	twi0_pins: twi0-pin-pd {
+		pins = "PD0", "PD12";
+		function = "i2c0";
+	};
+};
+
 &wdt {
         status = "okay";      
 };
+
 &usb_otg {
 	dr_mode = "otg";
 	status = "okay";
